auth/src/main/java/com/firebase/ui/auth/viewmodel/email/WelcomeBackPasswordHandler.java

```diff
@@ -61,38 +64,59 @@ public void startSignIn(@NonNull final String email,
+            AnonymousUpgradeUtils.validateCredential(credToValidate).addOnCompleteListener(
+                    new OnCompleteListener<AuthResult>() {
+                        @Override
+                        public void onComplete(@NonNull Task<AuthResult> task) {
+                            if (task.isSuccessful()) {
+                                handleMergeFailure(credToValidate);
+                            } else {
+                                setResult(Resource.<IdpResponse>forFailure(task.getException()));
+                            }
+                        }
+                    });
+
+        } else {
+            // Kick off the flow including signing in, linking accounts, and saving with SmartLock
+            getAuth().signInWithEmailAndPassword(email, password)
+                    .continueWithTask(new Continuation<AuthResult, Task<AuthResult>>() {
+                        @Override
+                        public Task<AuthResult> then(@NonNull Task<AuthResult> task) throws Exception {
+                            // Forward task failure by asking for result
+                            AuthResult result = task.getResult(Exception.class);
 
-                        // Task succeeded, link user if necessary
-                        if (credential == null) {
-                            return Tasks.forResult(result);
-                        } else {
-                            return result.getUser()
-                                    .linkWithCredential(credential)
-                                    .continueWithTask(new ProfileMerger(outputResponse))
-                                    .addOnFailureListener(new TaskFailureLogger(TAG, "linkWithCredential+merge failed."));
+                            // Task succeeded, link user if necessary
+                            if (credential == null) {
+                                return Tasks.forResult(result);
+                            } else {
+                                return result.getUser()
+                                        .linkWithCredential(credential)
+                                        .continueWithTask(new ProfileMerger(outputResponse))
+                                        .addOnFailureListener(new TaskFailureLogger(TAG,
+                                                "linkWithCredential+merge failed."));
+                            }
                         }
-                    }
-                })
-                .addOnCompleteListener(new OnCompleteListener<AuthResult>() {
-                    @Override
-                    public void onComplete(@NonNull Task<AuthResult> task) {
-                        if (!task.isSuccessful()) {
-                            setResult(Resource.<IdpResponse>forFailure(task.getException()));
-                            return;
+                    })
+                    .addOnCompleteListener(new OnCompleteListener<AuthResult>() {
```

Comment:
nit: fine either way but when your `onCompleteListener` just has a succcess-or-not branch you may find it clearer to use separate onSuccess and onFailure listeners.

Reply:
Done.